codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/protocols/ProtocolFunctions.kt

```diff
@@ -0,0 +1,162 @@
+     *     rust("fn $fnName(...) { ... }")
+     * }
+     * // Call the generated function
+     * rust("#T(...)", deserializeFn)
+     * ```
+     */
+    fun deserializeFn(
+        shape: Shape,
+        fnNameSuffix: String? = null,
+        block: ProtocolFnWritable,
+    ): RuntimeType = serDeFn(FnType.Deserialize, shape, serDeModule, block, fnNameSuffix)
+
+    private fun serDeFn(
+        fnType: FnType,
+        shape: Shape,
+        parentModule: RustModule.LeafModule,
+        block: ProtocolFnWritable,
+        fnNameSuffix: String? = null,
+    ): RuntimeType {
+        val moduleName = codegenContext.symbolProvider.shapeModuleName(codegenContext.serviceShape, shape)
+        val fnBaseName = codegenContext.symbolProvider.shapeFunctionName(codegenContext.serviceShape, shape)
+        val suffix = fnNameSuffix?.let { "_$it" } ?: ""
+        val fnName = RustReservedWords.escapeIfNeeded(
+            when (fnType) {
+                FnType.Deserialize -> "de_$fnBaseName$suffix"
+                FnType.Serialize -> "ser_$fnBaseName$suffix"
+            },
+        )
+        return serDeFn(moduleName, fnName, parentModule, block)
+    }
+
+    private fun serDeFn(
+        moduleName: String,
+        fnName: String,
+        parentModule: RustModule.LeafModule,
+        block: ProtocolFnWritable,
+    ): RuntimeType {
+        val additionalAttributes = when {
+            // Some SDK models have maps with names prefixed with `__mapOf__`, which become `__map_of__`,
+            // and the Rust compiler warning doesn't like multiple adjacent underscores.
+            moduleName.contains("__") || fnName.contains("__") -> listOf(Attribute.AllowNonSnakeCase)
+            else -> emptyList()
+        }
+        return RuntimeType.forInlineFun(
+            fnName,
+            RustModule.pubCrate(moduleName, parent = parentModule, additionalAttributes = additionalAttributes),
+        ) {
+            block(fnName)
+        }
+    }
+}
+
+/** Creates a module name for a ser/de function. */
+internal fun RustSymbolProvider.shapeModuleName(serviceShape: ServiceShape?, shape: Shape): String =
+    RustReservedWords.escapeIfNeeded(
+        "shape_" +
+            when (shape) {
+                is MemberShape -> model.expectShape(shape.container)
+                else -> shape
+            }.contextName(serviceShape).toSnakeCase(),
+        EscapeFor.ModuleName,
+    )
+
+/** Creates a unique name for a ser/de function. */
+internal fun RustSymbolProvider.shapeFunctionName(serviceShape: ServiceShape?, shape: Shape): String {
+    val containerName = when (shape) {
+        is MemberShape -> model.expectShape(shape.container).contextName(serviceShape).toSnakeCase()
+        else -> shape.contextName(serviceShape).toSnakeCase()
+    }
+    return when (shape) {
+        is MemberShape -> shape.memberName.toSnakeCase()
+        is DocumentShape -> "document"
+        else -> containerName
+    }
+}
+
+fun RustSymbolProvider.lensName(
```

Comment:
What is a lensName? Why the underscore prefix?

Reply:
I don't know much about lenses... just preserving functionality here.